File: terra-notification-service/src/services/eurekaClient.js
// services/eurekaClient.js - VERSION COMPLÈTEMENT REFACTORISÉE
import { Eureka } from 'eureka-js-client';

class EurekaService {
  constructor() {
    this.client = null;
    this.isRegistered = false;
    this.config = null;
  }

  initialize(config) {
    if (!config) {
      throw new Error('Configuration is required for Eureka initialization');
    }

    this.config = config;
    const { port = 4002, serviceName = 'terra-notification-service', eureka } = config;
    
    // IMPORTANT: Utiliser le nom Docker comme hostname
    const hostName = process.env.HOSTNAME || 'terra-notification-service';
    const eurekaHost = eureka?.host || process.env.EUREKA_HOST || 'terra-registry-service';
    const eurekaPort = eureka?.port || process.env.EUREKA_PORT || 8761;

    const timestamp = new Date().toISOString();
    const logger = {
      info: (msg) => console.log(`[${timestamp}] [Eureka] [INFO] ${msg}`),
      warn: (msg) => console.warn(`[${timestamp}] [Eureka] [WARN] ${msg}`),
      error: (msg) => console.error(`[${timestamp}] [Eureka] [ERROR] ${msg}`),
      debug: (msg) => console.debug(`[${timestamp}] [Eureka] [DEBUG] ${msg}`)
    };

    logger.info(`🎯 Initializing Eureka registration`);
    logger.info(`   Service: ${serviceName.toUpperCase()}`);
    logger.info(`   Port: ${port}`);
    logger.info(`   Hostname: ${hostName}`);
    logger.info(`   Eureka Server: ${eurekaHost}:${eurekaPort}`);

    try {
      this.client = new Eureka({
        // Configuration Eureka
        eureka: {
          host: eurekaHost,
          port: eurekaPort,
          servicePath: '/eureka/apps/',
          maxRetries: 10,
          requestRetryDelay: 2000,
          fetchRegistry: false,
          registerWithEureka: true,
        },
        
        // Configuration de l'instance
        instance: {
          app: 'TERRA-NOTIFICATION-SERVICE',
          hostName: hostName,
          ipAddr: hostName,
          instanceId: `${hostName}:${port}`,
          statusPageUrl: `http://${hostName}:${port}/health`,
          healthCheckUrl: `http://${hostName}:${port}/health`,
          homePageUrl: `http://${hostName}:${port}`,
          port: {
            '$': parseInt(port),
            '@enabled': true,
          },
          vipAddress: 'TERRA-NOTIFICATION-SERVICE',
          secureVipAddress: 'TERRA-NOTIFICATION-SERVICE',
          dataCenterInfo: {
            '@class': 'com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo',
            name: 'MyOwn',
          },
          metadata: {
            'instanceId': `${hostName}:${port}`,
            'version': process.env.APP_VERSION || '1.0.0',
            'service.type': 'notification'
          },
          leaseInfo: {
            renewalIntervalInSecs: 30,
            durationInSecs: 90
          }
        },
        
        // Logger amélioré
        logger: logger
      });

      // Événements avec logging amélioré
      this.client.on('started', () => {
        logger.info('🚀 Eureka client started');
      });

      this.client.on('registered', () => {
        this.isRegistered = true;
        logger.info(`✅ Successfully registered: ${serviceName.toUpperCase()} (${hostName}:${port})`);
      });

      this.client.on('deregistered', () => {
        this.isRegistered = false;
        logger.warn('🏁 Deregistered from Eureka');
      });

      this.client.on('heartbeat', () => {
        logger.debug('💓 Heartbeat sent to Eureka');
      });

      this.client.on('registryUpdated', () => {
        logger.debug('📋 Registry updated');
      });

      this.client.on('registryFetchFailed', (error) => {
        logger.warn(`⚠️ Registry fetch failed: ${error.message}`);
      });

      this.client.on('heartbeatFailed', (error) => {
        logger.error(`❌ Heartbeat failed: ${error.message}`);
      });

      logger.info('✅ Eureka client initialized successfully');
      return this.client;
      
    } catch (error) {
      logger.error(`💥 Failed to initialize Eureka client: ${error.message}`);
      logger.error(`Stack: ${error.stack}`);
      throw error;
    }
  }

  start() {
    if (!this.client) {
      if (!this.config) {
        console.error('[Eureka] ❌ Cannot start: Configuration not initialized');
        return;
      }
      this.initialize(this.config);
    }

    try {
      const timestamp = new Date().toISOString();
      console.log(`[${timestamp}] [Eureka] [INFO] Starting Eureka registration...`);
      
      this.client.start(error => {
        const timestamp = new Date().toISOString();
        if (error) {
          console.error(`[${timestamp}] [Eureka] [ERROR] Registration failed: ${error.message}`);
          console.error(`[${timestamp}] [Eureka] [ERROR] Stack: ${error.stack}`);
          console.log(`[${timestamp}] [Eureka] [INFO] Retrying in 10 seconds...`);
          setTimeout(() => this.start(), 10000);
        } else {
          console.log(`[${timestamp}] [Eureka] [INFO] Registration process started`);
        }
      });
      
    } catch (error) {
      const timestamp = new Date().toISOString();
      console.error(`[${timestamp}] [Eureka] [ERROR] Start error: ${error.message}`);
      console.error(`[${timestamp}] [Eureka] [ERROR] Stack: ${error.stack}`);
    }
  }

  stop() {
    if (this.client) {
      const timestamp = new Date().toISOString();
      console.log(`[${timestamp}] [Eureka] [INFO] Stopping Eureka client...`);
      try {
        this.client.stop();
        this.isRegistered = false;
        console.log(`[${timestamp}] [Eureka] [INFO] ✅ Eureka client stopped`);
      } catch (error) {
        console.error(`[${timestamp}] [Eureka] [ERROR] Error stopping client: ${error.message}`);
      }
    }
  }

  isConnected() {
    return this.isRegistered;
  }
}

export default new EurekaService();

File: terra-auth-service/auth_app/eureka_client.py
# eureka_client.py
import threading
import time
import os
import logging
import requests
from typing import Optional, Dict, Any
from django.conf import settings

logger = logging.getLogger(__name__)

class EurekaClient:
    def __init__(self, app_name: str, default_port: str = "8080"):
        self.app_name = app_name
        self.instance_id = None
        self.eureka_url = None
        self.heartbeat_thread = None
        self.running = threading.Event()
        self.default_port = default_port
        
    def get_config(self) -> Dict[str, Any]:
        """Récupère la configuration depuis Django settings ou variables d'environnement"""
        config = {
            "app_name": self.app_name,
            "port": os.getenv("SERVER_PORT", 
                             getattr(settings, "SERVER_PORT", self.default_port)),
            "hostname": os.getenv("EUREKA_INSTANCE_HOSTNAME", 
                                os.getenv("SERVICE_NAME", self.app_name.lower().replace("_", "-"))),
            "eureka_url": os.getenv("EUREKA_CLIENT_SERVICE_URL_DEFAULTZONE",
                                   getattr(settings, "EUREKA_URL", 
                                          "http://terra-registry-service:8761/eureka")),
            "health_path": os.getenv("HEALTH_CHECK_PATH", "/health"),
            "info_path": os.getenv("INFO_PATH", "/info")
        }
        return config
    
    def register_to_eureka(self) -> bool:
        """Enregistre l'instance auprès d'Eureka"""
        config = self.get_config()
        
        self.eureka_url = config["eureka_url"].rstrip("/")
        self.instance_id = f"{config['hostname']}:{config['port']}"
        
        payload = {
            "instance": {
                "instanceId": self.instance_id,
                "hostName": config["hostname"],
                "app": self.app_name.upper(),
                "ipAddr": config["hostname"],
                "port": {"$": int(config["port"]), "@enabled": True},
                "securePort": {"$": 443, "@enabled": False},
                "status": "UP",
                "healthCheckUrl": f"http://{config['hostname']}:{config['port']}{config['health_path']}",
                "statusPageUrl": f"http://{config['hostname']}:{config['port']}{config['info_path']}",
                "homePageUrl": f"http://{config['hostname']}:{config['port']}/",
                "dataCenterInfo": {
                    "@class": "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo",
                    "name": "MyOwn"
                },
                "metadata": {
                    "instanceId": self.instance_id,
                    "management.port": config["port"],
                    "version": os.getenv("APP_VERSION", "1.0.0"),
                    "zone": os.getenv("DEPLOYMENT_ZONE", "default")
                },
                "leaseInfo": {
                    "renewalIntervalInSecs": 30,
                    "durationInSecs": 90,
                    "registrationTimestamp": int(time.time() * 1000)
                },
                "vipAddress": config["app_name"].lower(),
                "secureVipAddress": config["app_name"].lower()
            }
        }
        
        try:
            max_retries = 3
            for attempt in range(max_retries):
                try:
                    resp = requests.post(
                        f"{self.eureka_url}/apps/{self.app_name.upper()}",
                        json=payload,
                        headers={"Content-Type": "application/json", "Accept": "application/json"},
                        timeout=15
                    )
                    
                    if resp.status_code in [200, 204]:
                        logger.info(f"[Eureka] ✅ {self.app_name} enregistré : {self.instance_id}")
                        return True
                    else:
                        logger.warning(f"[Eureka] Tentative {attempt+1}/{max_retries} échouée : {resp.status_code}")
                        if attempt < max_retries - 1:
                            time.sleep(5 * (attempt + 1))
                            
                except requests.exceptions.ConnectionError:
                    logger.warning(f"[Eureka] Eureka indisponible, tentative {attempt+1}/{max_retries}")
                    time.sleep(10 * (attempt + 1))
                    
            logger.error(f"[Eureka] ❌ Échec d'enregistrement après {max_retries} tentatives")
            return False
            
        except Exception as e:
            logger.error(f"[Eureka] Erreur d'enregistrement : {e}")
            return False
    
    def start_heartbeat(self):
        """Démarre le heartbeat périodique avec re-registration si nécessaire"""
        self.running.set()
        
        def heartbeat_loop():
            failed_attempts = 0
            max_failed_attempts = 3
            
            while self.running.is_set():
                try:
                    time.sleep(25)  # 5 secondes avant la deadline de 30s
                    
                    if not self.instance_id or not self.eureka_url:
                        logger.warning("[Eureka] Instance non enregistrée, tentative de re-registration...")
                        if self.register_to_eureka():
                            failed_attempts = 0
                        continue
                    
                    # Envoi du heartbeat
                    resp = requests.put(
                        f"{self.eureka_url}/apps/{self.app_name.upper()}/{self.instance_id}",
                        timeout=10
                    )
                    
                    if resp.status_code in [200, 204]:
                        failed_attempts = 0
                        logger.debug(f"[Eureka] ♥ Heartbeat réussi pour {self.instance_id}")
                    else:
                        failed_attempts += 1
                        logger.warning(f"[Eureka] Heartbeat échoué ({failed_attempts}/{max_failed_attempts})")
                        
                        if failed_attempts >= max_failed_attempts:
                            logger.warning("[Eureka] Trop d'échecs, re-registration...")
                            if self.register_to_eureka():
                                failed_attempts = 0
                            
                except requests.exceptions.RequestException as e:
                    failed_attempts += 1
                    logger.warning(f"[Eureka] Erreur heartbeat: {e}")
                    
                    if failed_attempts >= max_failed_attempts:
                        logger.warning("[Eureka] Tentative de re-registration après erreurs réseau")
                        if self.register_to_eureka():
                            failed_attempts = 0
                
                except Exception as e:
                    logger.error(f"[Eureka] Erreur inattendue dans heartbeat: {e}")
        
        self.heartbeat_thread = threading.Thread(target=heartbeat_loop, daemon=True)
        self.heartbeat_thread.start()
        logger.info(f"[Eureka] Heartbeat démarré pour {self.app_name}")
    
    def deregister(self):
        """Désenregistre l'instance d'Eureka"""
        self.running.clear()
        
        if self.heartbeat_thread:
            self.heartbeat_thread.join(timeout=5)
        
        if self.instance_id and self.eureka_url:
            try:
                requests.delete(f"{self.eureka_url}/apps/{self.app_name.upper()}/{self.instance_id}",
                              timeout=10)
                logger.info(f"[Eureka] Service {self.app_name} désenregistré")
            except Exception as e:
                logger.warning(f"[Eureka] Erreur lors du désenregistrement: {e}")
